### PR TITLE
feature(addon-google-analytics) utilise storybook addonParameter instead of window/global scope.

### PR DESCRIPTION
**Pull request by alexanderjeurissen**
## What I did

I changed the google analytics add-on to utilise `addonParameters` instead of globals.
This has several benefits:

- It makes the google analytics add-on more in line with other storybook add-on configurations.
- It opens up new ways to extend the configuration in future PR's, such as:
  - specifying GA option overrides for specific stories
  - allowing opt-in / opt-out of certain GA tracking events
  - Expanding the number of events that we allow to track such as: (search, keydown, control interaction etc.)

## How to test

- Is this testable with Jest or Chromatic screenshots?
  -  This is probably testable with Jest, although I could not find existing test coverage for this.
- Does this need a new example in the kitchen sink apps?
  - I don't think so, as we currently don't have analytics in the kitchen sink apps.
- Does this need an update to the documentation?
  - Yes, I updated the README to reflect the configuration change

### CODE REVIEW

**Comment by shilman** [comment]
@alexanderjeurissen thanks for the PR -- def a big improvement over the current version!

@ndelangen does this belong in addons.setConfig?

**Comment by alexanderjeurissen** [comment]
@shilman you're welcome, I'm currently working around some of these limitations more specifically the extensibility of the current google analytics package in our storybook at HackerOne.

This is a first step to get there, after this gets merged I want to explore expanding the CORE events that we track.

I also want to explore putting React GA in a provider so that we can also consume ReactGA in stories directly without having to re-initialize.

I recently authored http://npmjs.com/@alexanderjeurissen/use-reactga which could be a good fit here.

**Comment by alexanderjeurissen** [comment]
@shilman @ndelangen I'm not sure if this belongs in `addons.setConfig` for now it might make sense given that the configuration is global to Storybook. But if at a future point we want to enable users to override these settings per story, `addons.setConfig` might not make sense anymore ? unless there is a way to override `setConfig` per story.

**Comment by shilman** [comment]
@alexanderjeurissen i agree that parameters make more sense if we want to vary these things per story. is that something you have planned?

one other thing. this is a breaking change, so we should make sure to get it in before 6.0 goes out if we're going to do it. if you're planning on making more breaking changes in the near future, please consider that as well!

**Comment by alexanderjeurissen** [comment]
> @alexanderjeurissen i agree that parameters make more sense if we want to vary these things per story. is that something you have planned?

Yes I think it makes sense to allow for this given the minimal overhead that this introduces. I think that extending the current "global" parameters to allow for story specific ones is a non-breaking change as it only requires re-initializing a central `ReactGA` the API remains the same for global parameters.
> one other thing. this is a breaking change, so we should make sure to get it in before 6.0 goes out if we're going to do it. if you're planning on making more breaking changes in the near future, please consider that as well!

That makes sense, do you have an estimate as to when you expect a feature freeze for the 6.0 release to take effect ? That way I can plan some of these changes accordingly.

File: lib/cli/versions.json [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
+  "@storybook/marko": "6.0.0-beta.37",
+  "@storybook/mithril": "6.0.0-beta.37",
+  "@storybook/preact": "6.0.0-beta.37",
+  "@storybook/rax": "6.0.0-beta.37",
+  "@storybook/react": "6.0.0-beta.37",
+  "@storybook/riot": "6.0.0-beta.37",
+  "@storybook/server": "6.0.0-beta.37",
+  "@storybook/svelte": "6.0.0-beta.37",
+  "@storybook/vue": "6.0.0-beta.37",
+  "@storybook/web-components": "6.0.0-beta.37"
```
Comment (alexanderjeurissen) [inline lib/cli/versions.json:16]: Not sure where these changes are coming from 🤔

**Comment by alexanderjeurissen** [comment]
@shilman is there anything I need to do for the `Danger JS` failure ?

**Comment by shilman** [comment]
@alexanderjeurissen We’re scheduled to do the first RC this week. I think we can fudge it a bit for analytics because it’s not as widely used as some addons. We can also add this as a feature in 6.1 as needed as long as we provide back compat, which wouldn’t be hard in this case

**Comment by shilman** [comment]
Re: danger, don’t worry about it. It always fails on PRs from forks

**Comment by alexanderjeurissen** [comment]
@shilman I made some changes, I think that default assignment in the `addonParameter` destructuring will result in backwards compatibility.

Can you sanity check and if backwards compatibility looks good remove the `BREAKING CHANGE` label accordingly ?

**Comment by alexanderjeurissen** [comment]
I also created a followup PR that is based on this one (https://github.com/storybookjs/storybook/pull/11340) Let me know how you want to move forward with merging / releasing.

**Comment by alexanderjeurissen** [comment]
@shilman I respect that you and the other maintainers are currently very busy getting Storybook 6.0 ready to ship.

I wanted to inquire if there is any remaining action required on my end, or if this PR is ready to review as is.

**Comment by shilman** [comment]
@alexanderjeurissen Looking good AFAIK. I’ll try to review today

File: addons/google-analytics/README.md [hunk 2]
```diff
@@ -20,9 +26,25 @@ module.exports = {
...
 
+```js
+import { addParameters } from '@storybook/client-api';
+
+addParameters({
+  analytics: {
+    reactGAId: 'UA-000000-01',
+    reactGAOptions: {},
+  },
+});
```
Comment (ndelangen) [inline addons/google-analytics/README.md:42]: setConfig would make more sense.

Even more would defining this in `main.js`, and have a preset push it into the proper place where the runtime expects it. (such as setConfig)

Comment (alexanderjeurissen) [reply addons/google-analytics/README.md:42]: @ndelangen can you elaborate why `setConfig` would be better in this scenario ?

If I understand correctly, `setConfig` does not allow for local story overrides. I can think of future enhancements where we would want to allow users to specify different reactGA options or tracking codes for certain stories. Would that still work with a setConfig approach ?

Comment (ndelangen) [reply addons/google-analytics/README.md:42]: > allow users to specify different reactGA options or tracking codes for certain stories

I suspect there's no need for that.. Do you?

Comment (alexanderjeurissen) [reply addons/google-analytics/README.md:42]: I do, having different tracking codes is likely desired in cases where a medium to big UI library is published to different scoped NPM packages but hosted in a single repository.

for instance say we have a UI kit that is split up across the following NPM packages:

`@same_fancy_library/components`
`@some_fancy_library/utils`

This pattern is getting more and more popular, in those cases one could argue having separate storybook instances for each package, but from my experience having a single storybook instance, that hosts all stories is preferable from a end user UX perspective.

Furthermore I've seen instances where submodules of big component libraries are maintained by different teams who each have their own GA View / property that they want to observe.

Regardless if we want to optimise for this use-case, I think that the overhead of enabling this use-case is so small that it is likely worth exploring.

Comment (alexanderjeurissen) [reply addons/google-analytics/README.md:42]: @ndelangen WDYT ?

Comment (alexanderjeurissen) [reply addons/google-analytics/README.md:42]: I gave this some more thought, and given that storybook will get composition soon (https://github.com/storybookjs/storybook/pull/9210) my point is moot.

I'll look into updating this PR to use `setConfig`

Comment (ndelangen) [reply addons/google-analytics/README.md:42]: 👍

**Pull request closed without merging**